---
output: html_document
title: "Comparative analysis of Alignment/Mapping Pipelines in SEQC dataset"
author: "Mira Sohn" 
---


```{r global_options, include=FALSE}


knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE
    )


```

## Loading packages

```{r loading_packages}

library(data.table)
library(tidyverse)
library(rmarkdown)
library(ggplot2)
library(pheatmap)
library(ggrepel)
library(gridExtra)
library(AnnotationHub)
library(tximport)
library(Rsubread)
library(DESeq2)


```

## Setting AnnotationHub
### Assign your species of interest


```{r annotationhub_setup}


AnnotationSpecies <- "Homo sapiens"  # Assign your species 
ah <- AnnotationHub(hub=getAnnotationHubOption("URL"))   # Bring annotation DB

```



## Running AnnotationHub

```{r running_annotationhub}

# Filter annotation of interest
ahQuery <- query(ah, c("OrgDb", AnnotationSpecies))      
if (length(ahQuery) == 1) {
    DBName <- names(ahQuery)
} else if (length(ahQuery) > 1) {
               DBName <- names(ahQuery)[1]
} else {
    print("You don't have a valid DB")
    rmarkdown::render() 
} 


AnnoDb <- ah[[DBName]] # Store into an OrgDb object  


# Explore your OrgDb object with following accessors:
# columns(AnnpDb)
# keytypes(AnnoDb)
# keys(AnnoDb, keytype=..)
# select(AnnoDb, keys=.., columns=.., keytype=...)
AnnoKey <- keys(AnnoDb, keytype="ENSEMBLTRANS")

AnnoDb <- select(AnnoDb, 
                 AnnoKey,
                 keytype="ENSEMBLTRANS",  # genes
                 columns=c("ENSEMBL", "SYMBOL"))



head(AnnoDb)
```

## Checking out the AnnotationHub output 

```{r checking_annotationhub_output}

# Check if your AnnoDb has been extracted and saved correctely
class(AnnoDb)
head(AnnoDb)

```


## Metadata setting

```{r}
# This code chunk needs to be written by yourself 

# Define sample names 
SampleNames <-  c(paste0("A", 1:3), paste0("B", 1:3)) 

# Aligner names
Alingers <- c("Salmon", "STAR", "HISAT2")

# Define group level
GroupLevel <- c("A", "B")

# Define contrast for DE analysis
Contrast <- c("Group", GroupLevel)

# Set a function for file paths
path.fn <- function(head, tail) { 

    vec <- c(paste0(head,    # head = e.g. "hisat2", "star", or "salmon"
                    "_output/",
                    SampleNames,
                    tail))   # tail = file name after SampleNames 
}


# Define .sf file path
sf <- path.fn("salmon",".salmon_quant/quant.sf")


# Define STAR file path
star <- path.fn("star", "Aligned.sortedByCoord.out.bam")

# Define HISAT2 file path
hisat <- path.fn("hisat2", ".sorted.bam")


# Define sample groups
group <- c(rep("A", 3), rep("B", 3))

# Create metadata
metadata <- data.frame(Sample=factor(SampleNames, levels=SampleNames),
                       Group=factor(group, levels=GroupLevel),
                       Salmon_path=sf,
                       STAR_path=star, 
                       HISAT2_path=hisat)

# Assign row names with sample names
rownames(metadata) <- SampleNames


# Explore the metadata
print(metadata)


```

## featureCounts parameter setting

```{r featurecounts_parameters}

# "mm10", "mm9", "hg38", or "hg19"
annot.inbuilt <- "hg38"

# GTF file path
annot.ext <- "reference_GENCODE/gencode.v36.primary_assembly.annotation.gtf"

# annotation type:
# e.g.: "gene_id", "transcript_id", or "gene_name"
GTF.attrType <- "gene_name"

# Number of cores 
nthread <- 8

# Set a function importing counts from BAM files with featureCounts()
fcounts.fn <- function(vec) {

    fc <- featureCounts(files=vec,   # a vector assigning BAM file paths
                         annot.inbuilt=annot.inbuilt,
                         annot.ext=annot.ext,
                         GTF.attrType=GTF.attrType,
                         isGTFAnnotationFile=T,
                         nthread=nthread, 
                         isPairedEnd=T,
                         verbose=T)

    return(fc$counts)
}


```

## Importing counts 

### Importing Salmon counts

#### Note:     

#### txi1 <- tximport(..., txOut=F)   
#### txi2 <- tximport(..., txOut=T)     
#### txi2 <- summarizedToGene(...)    

#### counts extracted from txi1 and txi2 are the same 


```{r import_salmon}

# Import gene level summarized counts 
salmon.txi <- tximport(metadata$Salmon_path, 
                       type = "salmon", 
                       tx2gene=AnnoDb[, -3],   # Annotate at gene level (ENSEMBL)
                       ignoreTxVersion=T) 

# Extract the counts and save as a data frame
salmon.counts <- salmon_txi$counts

# Explore the salmon count data frame
head(salmon.counts)
dim(salmon.counts)
summary(salmon.counts)
```

### Importing STAR counts


```{r import_star}

# Extract counts by running featureCounts() 
star.counts <- fcounts.fn(metadata$STAR_path)


# Explore the STAR count data frame
head(star.counts)
dim(star.counts)
summary(star.counts)

```


### Importing HISAT2 counts

```{r import_hisat2}

# Extract counts by running featureCounts() 
hisat2.counts <- fcounts.fn(metadata$HISAT2_path)

# Explore the HISAT2 count data frame
head(hisat2.counts)
dim(hisat2.counts)
summary(hisat2.counts)

```

## Data cleaning: new annotation and row naming 


```{r count_dataframe_cleaning}

countList <- list(Aligners[1]=salmon.counts, 
                  Alingers[2]=star.counts,
                  Aligners[3]=hisat2.counts)


```


```{r session_info}

sessionInfo()

```

